# **Shark Attack Data Analysis**

# Exploring Global Shark–Human Interaction Patterns

## Objectives

Project objectives:
- Analyse global shark attack incident data.
- Identify patterns in shark–human interactions.
- Explore geographic and behavioural factors associated with shark attacks.
- Develop an interactive dashboard communicating insights clearly.
- Evaluate ethical considerations related to wildlife incident data.
- Document the full data analytics project lifecycle.

Notebook objectives:
- Load and inspect the global shark attacks dataset.
- Clean and anonymise the dataset by removing personally identifiable information.
- Explore patterns in shark incidents across time, geography, and human activities.
- Analyse the relationship between activity type and fatality outcomes.
- Generate insights that will be used to build an interactive Power BI dashboard.


## Inputs

- data/raw/shark_attacks.csv – raw shark attack dataset obtained from Kaggle.
- Python libraries including pandas, numpy, matplotlib, seaborn, and plotly.
- Jupyter Notebook environment for data analysis and visualisation.


## Outputs

- Cleaned dataset for analysis.
- Exploratory data visualisations.
- Summary statistics describing shark attack patterns.
- Processed dataset exported for use in the Power BI dashboard.
- Visual outputs saved in outputs/figures.


## Additional Comments

- The dataset contains historical records of shark attack incidents from multiple countries.
- Personal identifiers such as victim names will be removed to ensure ethical data handling.
- The analysis focuses on identifying patterns in time, location, activity type, and fatality outcomes.
- Findings from this notebook will support the development of an interactive dashboard.



---

# Change working directory

* We are assuming you will store the notebooks in a subfolder, therefore when running the notebook in the editor, you will need to change the working directory

We need to change the working directory from its current folder to its parent folder
* We access the current directory with os.getcwd()

In [1]:
import os
current_dir = os.getcwd()
current_dir

'c:\\Users\\Shazia Mujahid\\Documents\\shark-attack-project\\shark-attack-data-dashboard\\jupyter_notebooks'

We want to make the parent of the current directory the new current directory
* os.path.dirname() gets the parent directory
* os.chir() defines the new current directory

In [2]:
os.chdir(os.path.dirname(current_dir))
print("You set a new current directory")

You set a new current directory


Confirm the new current directory

In [3]:
current_dir = os.getcwd()
current_dir

'c:\\Users\\Shazia Mujahid\\Documents\\shark-attack-project\\shark-attack-data-dashboard'

# 1. Dataset Overview and Loading

In this section, the shark attack dataset is loaded and inspected to understand the structure. Initial checks are performed to examine the number of observations, available variables, missing values, and general dataset characteristics before moving onto the data cleaning stage. This step is crucial for identifying potential data quality issues before the data cleaning stage.


In [4]:
# Import libraries for data analysis and visualisation 

import pandas as pd     # Used for data manipulation and analysis
import numpy as np      # Provides numerical operations and array support
import matplotlib.pyplot as plt   # Used for creating basic visualisations
import seaborn as sns   # Built on matplotlib for improved statistical visualisations

In [5]:
# Load the raw shark attack dataset

df = pd.read_csv("data/raw/shark_attacks.csv")

# Display the first few rows to preview the dataset structure

df.head()

,date,year,type,country,area,location,activity,name,sex,age,fatal_y_n,time,species
0,2023-05-13,2023.0,Unprovoked,AUSTRALIA,South Australia,Elliston,Surfing,Simon Baccanello,M,46,Y,10h10,White shark
1,2023-04-29,2023.0,Unprovoked,AUSTRALIA,Western Australia,"Yallingup, Busselton",Swimming,male,M,NaN,N,11h20,1m shark
2,2022-10-07,2022.0,Unprovoked,AUSTRALIA,Western Australia,Port Hedland,Spearfishing,Robbie Peck,M,38,N,11h30,Bull shark
3,2021-10-04,2021.0,Unprovoked,USA,Florida,"Fort Pierce State Park, St. Lucie County",Surfing,Truman Van Patrick,M,25.0,N,NaN,NaN
4,2021-10-03,2021.0,Unprovoked,USA,Florida,"Jensen Beach, Martin County",Swimming,male,M,NaN,N,12h00,NaN


In [6]:
# Examine the size of the dataset
# This shows the total number of rows (records) and columns (variables)

df.shape

(6890, 13)

The dataset contains 6,890 recorded shark incident observations and 13 variables describing each incident. Each row represents a single shark attack record, while the columns contain information such as the date, location, activity being performed, shark species involved, and whether the incident resulted in a fatality.

In [7]:
# Display the column names to understand which variables are available
# This helps identify which variables may be relevant for analysis

df.columns

Index(['date', 'year', 'type', 'country', 'area', 'location', 'activity',
       'name', 'sex', 'age', 'fatal_y_n', 'time', 'species'],
      dtype='object')

The dataset contains 13 variables describing each shark incident. These variables include temporal information such as the date and year of the incident, geographic information including country and location, and contextual details such as the activity being performed and the shark species involved.

The dataset also contains demographic attributes of individuals involved in the incidents and an indicator of whether the incident resulted in a fatality.

In [8]:
# Display dataset information including data types and non-null values
# This helps identify missing data and potential type conversions required during cleaning

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6890 entries, 0 to 6889
Data columns (total 13 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   date       6587 non-null   object 
 1   year       6758 non-null   float64
 2   type       6871 non-null   object 
 3   country    6839 non-null   object 
 4   area       6409 non-null   object 
 5   location   6325 non-null   object 
 6   activity   6304 non-null   object 
 7   name       6670 non-null   object 
 8   sex        6318 non-null   object 
 9   age        3903 non-null   object 
 10  fatal_y_n  6890 non-null   object 
 11  time       3372 non-null   object 
 12  species    3772 non-null   object 
dtypes: float64(1), object(12)
memory usage: 699.9+ KB


The dataset contains 6,890 observations and 13 variables describing shark incident records. Most variables are stored as object (string) data types, while the year variable is stored as a numeric value (float64).

Several variables contain missing values, which will need to be addressed during the data cleaning stage. For example, the variables age, time, and species contain a large number of missing entries.

The variable name contains personally identifiable information and will be removed during the cleaning process to ensure the dataset is anonymised before analysis.

In [9]:
# Check the number of missing values in each column
# Missing data may require cleaning or transformation before analysis

df.isnull().sum()

date          303
year          132
type           19
country        51
area          481
location      565
activity      586
name          220
sex           572
age          2987
fatal_y_n       0
time         3518
species      3118
dtype: int64

The missing value analysis shows that several variables in the dataset contain incomplete information. Some variables such as type, country, and year contain relatively small numbers of missing values and can be retained with minimal cleaning.

However, other variables contain substantial missing data. For example, the age, time, and species columns contain a large proportion of missing values. These variables may require careful handling or may be excluded from certain analyses depending on their completeness.

As previously mentioned, the name column contains personally identifiable information and will be removed during the cleaning stage to ensure the dataset used for analysis is anonymised.

---

# 2. Data Preparation

This section prepares the dataset for further analysis. Initial preparation steps are performed to ensure the dataset is suitable for cleaning and analysis.

These steps include removing variables that are not required for the analysis, preparing column names for easier manipulation, and converting certain variables into more appropriate data types. Preparing the dataset in this way helps ensure consistency and improves the efficiency of subsequent data cleaning and feature engineering steps.

## 2.1 Remove Personally Identifiable Information

The dataset contains a column labelled `name`, which includes the names of individuals involved in shark attack incidents. Since this information is not required for the analysis and represents personally identifiable information, the column is removed to maintain data privacy and ensure the dataset used for analysis is anonymised.

In [10]:
# Remove the 'name' column from the dataset.
# This column contains personally identifiable information (PII), which is not needed for the analysis and should be excluded to help preserve privacy and anonymise the dataset.

df = df.drop(columns=["name"])

# Display the first few rows to confirm that the column has been removed from the DataFrame.

df.head()

,date,year,type,country,area,location,activity,sex,age,fatal_y_n,time,species
0,2023-05-13,2023.0,Unprovoked,AUSTRALIA,South Australia,Elliston,Surfing,M,46,Y,10h10,White shark
1,2023-04-29,2023.0,Unprovoked,AUSTRALIA,Western Australia,"Yallingup, Busselton",Swimming,M,NaN,N,11h20,1m shark
2,2022-10-07,2022.0,Unprovoked,AUSTRALIA,Western Australia,Port Hedland,Spearfishing,M,38,N,11h30,Bull shark
3,2021-10-04,2021.0,Unprovoked,USA,Florida,"Fort Pierce State Park, St. Lucie County",Surfing,M,25.0,N,NaN,NaN
4,2021-10-03,2021.0,Unprovoked,USA,Florida,"Jensen Beach, Martin County",Swimming,M,NaN,N,12h00,NaN


## 2.2 Standardise Column Names

To improve readability and ensure consistency when referencing variables in code, column names are standardised by converting them to lowercase and removing any unnecessary whitespace.

In [11]:
# Standardise all column names by:
# 1. converting them to lowercase
# 2. removing leading and trailing whitespace

# This makes the dataset easier to work with and reduces the risk of coding errors caused by inconsistent naming conventions.

df.columns = df.columns.str.lower().str.strip()

# Display the updated column names to verify that the formatting has been applied correctly.
df.columns

Index(['date', 'year', 'type', 'country', 'area', 'location', 'activity',
       'sex', 'age', 'fatal_y_n', 'time', 'species'],
      dtype='object')

## 2.3 Convert the Year Variable

The `year` variable is currently stored as a floating-point data type. Since years are whole-number values, this variable is converted to numeric format to ensure consistency and support time-based analysis.

In [13]:
# Convert the 'year' column to a numeric data type.
# Any non-numeric values will be converted to NaN using errors='coerce'.

df["year"] = pd.to_numeric(df["year"], errors="coerce")

# Convert the column to pandas nullable integer type (Int64).
# This allows integer values while still supporting missing values.

df["year"] = df["year"].astype("Int64")

# Check the data type to confirm the conversion.

df["year"].dtype

Int64Dtype()

## 2.4 Convert the Date Variable

The `date` column is converted from text format to a datetime format. This ensures the variable is suitable for chronological analysis and allows dates to be handled more efficiently in pandas.

In [14]:
# Convert the 'date' column to datetime format.
# Invalid or inconsistent date values will be converted to NaT (Not a Time) using errors='coerce'.

# This allows the date variable to be used for time-based analysis, filtering, and sorting.

df["date"] = pd.to_datetime(df["date"], errors="coerce")

# Display the first few values from the 'date' column to verify that the conversion was successful.

df["date"].head()

0   2023-05-13
1   2023-04-29
2   2022-10-07
3   2021-10-04
4   2021-10-03
Name: date, dtype: datetime64[ns]

## 2.5 Verify Dataset Structure

After the initial preparation steps, the dataset structure is reviewed to confirm that the transformations have been applied correctly and that the dataset is ready for the cleaning stage.

In [15]:
# Display a summary of the dataset structure after preparation.
# This includes:
# - the number of rows and columns
# - column names
# - non-null counts
# - data types

# This step is used to confirm that:
# - the 'name' column has been removed
# - the 'date' column is now datetime
# - the 'year' column has been converted to numeric format

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6890 entries, 0 to 6889
Data columns (total 12 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   date       6577 non-null   datetime64[ns]
 1   year       6758 non-null   Int64         
 2   type       6871 non-null   object        
 3   country    6839 non-null   object        
 4   area       6409 non-null   object        
 5   location   6325 non-null   object        
 6   activity   6304 non-null   object        
 7   sex        6318 non-null   object        
 8   age        3903 non-null   object        
 9   fatal_y_n  6890 non-null   object        
 10  time       3372 non-null   object        
 11  species    3772 non-null   object        
dtypes: Int64(1), datetime64[ns](1), object(10)
memory usage: 652.8+ KB


### Section 2 Summary

The dataset has now undergone initial preparation steps. Personally identifiable information has been removed, and key variables such as `date` and `year` have been formatted to support analysis.

These preparation steps improve dataset consistency and ensure that the data is ready for the next stage, which focuses on cleaning missing values, resolving inconsistencies, and engineering features for analysis.

---

NOTE

* You may add as many sections as you want, as long as it supports your project workflow.
* All notebook's cells should be run top-down (you can't create a dynamic wherein a given point you need to go back to a previous cell to execute some task, like go back to a previous cell and refresh a variable content)

---

# Push files to Repo

* In cases where you don't need to push files to Repo, you may replace this section with "Conclusions and Next Steps" and state your conclusions and next steps.